src/cmd/sign.js

```diff
@@ -28,39 +30,105 @@ export default function sign(
           }
         })
         .then((manifestData) => {
-          if (!manifestData.applications) {
-            // TODO: remove this when signing supports manifests
-            // without IDs: https://github.com/mozilla/web-ext/issues/178
-            throw new InvalidManifest(
-              'applications.gecko.id in manifest.json is required for signing');
+          return Promise.all([
+            build({sourceDir, artifactsDir: tmpDir.path()},
+                  {manifestData}),
+            getIdFromArtifacts(artifactsDir),
+          ])
+          .then(([buildResult, idFromArtifacts]) => {
+            return {buildResult, manifestData, idFromArtifacts};
+          });
+        })
+        .then(({buildResult, manifestData, idFromArtifacts}) => {
+          const manifestId = getManifestId(manifestData);
+          if (id && manifestId) {
+            throw new WebExtError(
+              `Cannot set custom ID ${id} because manifest.json ` +
+              `declares ID ${manifestId}`);
+          }
+          if (manifestId) {
+            id = manifestId;
+          }
+          if (!id && idFromArtifacts) {
+            log.info(
+              'Using previously auto-generated extension ID: ' +
+              `${idFromArtifacts}`);
+            id = idFromArtifacts;
+          }
+          if (!id) {
+            log.warn('No extension ID specified (it will be auto-generated)');
           }
-          return manifestData;
+          return signAddon({
+            apiKey,
+            apiSecret,
+            apiUrlPrefix,
+            timeout,
+            verbose,
+            id,
+            xpiPath: buildResult.extensionPath,
+            version: manifestData.version,
+            downloadDir: artifactsDir,
+          });
         })
-        .then((manifestData) => {
-          return build(
-            {sourceDir, artifactsDir: tmpDir.path()},
-            {manifestData})
-            .then((buildResult) => {
-              return {buildResult, manifestData};
-            });
+        .then((signingResult) => {
+          if (signingResult.id) {
+            return saveIdToArtifacts(artifactsDir, signingResult.id)
+              .then(() => signingResult);
+          } else {
+            return signingResult;
+          }
         })
-        .then(({buildResult, manifestData}) => signAddon({
-          apiKey,
-          apiSecret,
-          apiUrlPrefix,
-          timeout,
-          verbose,
-          xpiPath: buildResult.extensionPath,
-          id: manifestData.applications.gecko.id,
-          version: manifestData.version,
-          downloadDir: artifactsDir,
-        }))
         .then((signingResult) => {
           // All information about the downloaded files would have
           // already been logged by signAddon().
-          log.info(signingResult.success ? 'SUCCESS' : 'FAIL');
+          if (signingResult.success) {
+            log.info(`Extension ID: ${signingResult.id}`);
```

Comment:
@kumar303 at line 74, it seems to me that in some cases the `id` property can be missing from the `signingResult` (my guess is that it could happen when the extension id is assigned statically in the `manifest.json`), and in that case the above line is going to print a "Extension ID: undefined" log, which can be confusing for the the user.

If the above assumptions are true, we can probably tweak it a bit, to ensure that it is going to always print a log which contains the `id` (even if it is the same statically assigned in the manifest.json)